rust/arrow/src/compute/kernels/take.rs

```diff
@@ -0,0 +1,519 @@
+                .map(|a| take(a, index, Some(&options)))
+                .collect();
+            let arrays = arrays?;
+            let pairs: Vec<(Field, ArrayRef)> =
+                fields.clone().into_iter().zip(arrays).collect();
+            Ok(Arc::new(StructArray::from(pairs)) as ArrayRef)
+        }
+        t @ _ => unimplemented!("Sort not supported for data type {:?}", t),
+    }
+}
+
+/// Options that define how `take` should behave
+#[derive(Clone)]
+pub struct TakeOptions {
+    /// perform bounds check before taking
+    pub check_bounds: bool,
+}
+
+impl Default for TakeOptions {
+    fn default() -> Self {
+        Self {
+            check_bounds: false,
+        }
+    }
+}
+
+/// `take` implementation for numeric arrays
+fn take_numeric<T>(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef>
+where
+    T: ArrowNumericType,
+{
+    let mut builder = PrimitiveBuilder::<T>::new(index.len());
+    let a = array.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append_null()?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_valid(ix) {
+                builder.append_value(a.value(ix))?;
+            } else {
+                builder.append_null()?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for binary arrays
+fn take_binary(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
+    let mut builder = BinaryBuilder::new(index.len());
+    let a = array.as_any().downcast_ref::<BinaryArray>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append(false)?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_null(ix) {
+                builder.append(false)?;
+            } else {
+                builder.append_values(a.value(ix))?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for boolean arrays
+fn take_bool(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
+    let mut builder = BooleanBuilder::new(index.len());
+    let a = array.as_any().downcast_ref::<BooleanArray>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append_null()?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_null(ix) {
+                builder.append_null()?;
+            } else {
+                builder.append_value(a.value(ix))?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for list arrays
+///
+/// Calculates the index and indexed offset for the inner array,
+/// applying `take` on the inner array, then reconstructing a list array
+/// with the indexed offsets
+fn take_list(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
```

Comment:
Some optimizations can be done here such as if it is taking the whole list or a contiguous sublist. This can be left as a TODO though.